adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,712 @@
+                result = 0;
+            }
+        }
+    }
+
+    return result;
+}
+
+
+/* Codes_SRS_TLSIO_SSL_ESP8266_99_013: [ The tlsio_openssl_close succeed.]*/
+int tlsio_openssl_close(CONCRETE_IO_HANDLE tls_io, ON_IO_CLOSE_COMPLETE on_io_close_complete, void* callback_context)
+{
+    int result = 0;
+
+    /* Codes_SRS_TLSIO_SSL_ESP8266_99_011: [ The tlsio_openssl_close NULL parameter.]*/
+    if (tls_io == NULL)
+    {
+        result = __LINE__;
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+
+        if ((tls_io_instance->tlsio_state == TLSIO_STATE_NOT_OPEN) ||
+            (tls_io_instance->tlsio_state == TLSIO_STATE_CLOSING))
+        {
+            /* Codes_SRS_TLSIO_SSL_ESP8266_99_012: [ The tlsio_openssl_close wrong state.]*/
+            result = __LINE__;
+            LogError("Invalid tlsio_state. Expected state is TLSIO_STATE_NOT_OPEN or TLSIO_STATE_CLOSING.");
+        }
+        else
+        {
+            tls_io_instance->tlsio_state = TLSIO_STATE_CLOSING;
+            tls_io_instance->on_io_close_complete = on_io_close_complete;
+            tls_io_instance->on_io_close_complete_context = callback_context;
+
+            destroy_openssl_instance(tls_io_instance);
+            result = 0;
+        }
+    }
+
+    return result;
+}
+
+int tlsio_openssl_send(CONCRETE_IO_HANDLE tls_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
+{
+    int result;
+
+    if (tls_io == NULL)
+    {
+        /* Codes_SRS_TLSIO_SSL_ESP8266_99_014: [ The tlsio_openssl_send NULL instance.]*/
+        result = __LINE__;
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+
+        if (tls_io_instance->tlsio_state != TLSIO_STATE_OPEN)
+        {
+            /* Codes_SRS_TLSIO_SSL_ESP8266_99_015: [ The tlsio_openssl_send wrog state.]*/
+            result = __LINE__;
+            LogError("Invalid tlsio_state. Expected state is TLSIO_STATE_OPEN.");
+        }
+        else
+        {
+            int total_write = 0;
+            int res = 0;
+            int retry = 0;
+
+            while(size > 0 && retry < 500){
+                /* Codes_SRS_TLSIO_SSL_ESP8266_99_016: [ The tlsio_openssl_send SSL_write success]*/
+                /* Codes_SRS_TLSIO_SSL_ESP8266_99_017: [ The tlsio_openssl_send SSL_write failure]*/
+                res = SSL_write(tls_io_instance->ssl, ((uint8*)buffer)+total_write, size);
+                LogInfo("SSL_write res: %d, size: %d, retry: %d", res, size, retry);
+                if(res > 0){
+                    total_write += res;
+                    size = size - res;
+                }
+                else
+                {
+                    retry++;
+                }
+            }
+
+            if (retry >= 500)
+            {
+                result = 1;
```

Comment:
Please call on_send_complete with IO_SEND_ERROR.

As a standard result should receive __LINE__.